rholang/src/main/scala/coop/rchain/rholang/interpreter/sort.scala

```diff
@@ -0,0 +1,311 @@
+/**
+  * Sorts the insides of the Par and ESet/EMap of the rholangADT
+  *
+  * A score tree is recursively built for each term and is used to sort the insides of Par/ESet/EMap.
+  * For most terms, the current term type's absolute value based on the Score object is added as a Leaf
+  * to the left most branch and the score tree built for the inside terms are added to the right.
+  * The Score object is a container of constants that arbitrarily assigns absolute values to term types.
+  * The sort order is total as every term type is assigned an unique value in the Score object.
+  * For ground types, the appropriate integer representation is used as the base score tree.
+  * For var types, the Debruijn level from the normalization is used.
+  *
+  * In order to sort an term, call [Type]SortMatcher.sortMatch(term)
+  * and extract the .term  of the returned ScoredTerm.
+  */
+
+package coop.rchain.rholang.interpreter
+
+import coop.rchain.rholang.intepreter._
+
+sealed trait Tree[T] {
+  def size: Int
+}
+case class Leaf[T](item: T) extends Tree[T] {
+  def size = 1
+}
+case class Node[T] (children: Seq[Tree[T]]) extends Tree[T] {
+  def size = children.map(_.size).sum
+}
+
+object Leaves {
+  // Shortcut to be able to write Leaves(1,2,3) instead of Node(Seq(Leaf(1),Leaf(2),Leaf(3)))
+  def apply[T](children: T*) : Node[T] = Node(children.map(a => Leaf(a)))
+}
+
+// Effectively a tuple that groups the term to its score tree.
+case class ScoredTerm[T](term: T, score: Tree[Int]) extends Ordered[ScoredTerm[T]] {
+  def compare(that: ScoredTerm[T]) : Int = {
+    def compareScore(s1: Tree[Int], s2: Tree[Int]) : Int = {
+      (s1, s2) match {
+        case (Leaf(a), Leaf(b)) => a.compare(b)
```

Comment:
Make the tree a tree of Either[Int, String]. Make sure there's a sort order for Either. If not, create your own case class and add a sort order.

Reply:
Done